File: kdb-storage-engine/src/jvmTest/kotlin/dev/kdb/storage/engine/WriteThroughputBenchTest.kt
package dev.kdb.storage.engine

import dev.kdb.storage.StorageEngineConfig
import dev.kdb.storage.io.FileBackedPlatformIoShimFactory
import dev.kdb.storage.io.PlatformIoConfig
import dev.kdb.storage.wal.DefaultWriteAheadLogFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import kotlin.io.path.createTempDirectory
import kotlin.test.Test
import kotlin.time.measureTime

/**
 * Direct engine-level write throughput measurement, mirroring
 * go/kdb/storage/engine/write_throughput_bench_test.go. Not run as part
 * of the default test task's assertions (it prints results and asserts
 * nothing) - a lightweight, bounded substitute for a full JMH benchmark
 * so real hardware concurrency (Dispatchers.Default, real OS threads) is
 * exercised without JMH's fork/warmup overhead. See
 * docs/benchmarks/phase0-baseline.md for the Go-side numbers this is
 * meant to be compared against.
 */
class WriteThroughputBenchTest {
    @Test
    fun writeBlobThroughput_diskWal() =
        runBlocking {
            for (parallelism in listOf(1, 8, 64, 256)) {
                val root = createTempDirectory("kdb-bench-wal-kt").toString()
                val shim = FileBackedPlatformIoShimFactory.open(PlatformIoConfig(rootDirectory = root))
                val config = StorageEngineConfig(globalMemoryBudgetBytes = 64L shl 20, ioShim = shim)
                val wal = DefaultWriteAheadLogFactory().openOrCreate("bench-ns", config, shim)
                val engine = ServerStorageEngine("bench-ns", config, wal)

                val opsPerWorker = 2000 / parallelism.coerceAtMost(2000)
                val totalOps = opsPerWorker * parallelism

                val elapsed =
                    measureTime {
                        coroutineScope {
                            (0 until parallelism).map { worker ->
                                async(Dispatchers.Default) {
                                    val payload = ByteArray(128)
                                    for (i in 0 until opsPerWorker) {
                                        payload[0] = (worker + i).toByte()
                                        engine.writeBlob(payload)
                                    }
                                }
                            }.forEach { it.await() }
                        }
                    }

                val nsPerOp = elapsed.inWholeNanoseconds / totalOps.coerceAtLeast(1)
                val opsPerSec = if (nsPerOp == 0L) 0 else 1_000_000_000L / nsPerOp
                println(
                    "parallel=$parallelism totalOps=$totalOps elapsed=$elapsed " +
                        "nsPerOp=$nsPerOp opsPerSec=$opsPerSec",
                )
                for (s in StageRecorder.Default.snapshot()) {
                    println(
                        "  stage=${s.stage} count=${s.count} mean=${s.mean} p50=${s.p50} p99=${s.p99} max=${s.max}",
                    )
                }
                StageRecorder.Default.reset()
            }
        }
}

File: go/kdb/storage/engine/write_throughput_bench_test.go
package engine_test

// Phase 0 baseline benchmark for the write path described in
// docs/benchmarks/phase0-baseline.md. It measures ServerEngine.WriteBlob
// throughput under concurrent load against a real, disk-backed WAL, so
// the fsync cost is representative rather than optimistic in-memory
// numbers. See metrics.Default for the accompanying per-stage
// (lock_wait / fsync_wait) breakdown printed after each run.

import (
	"fmt"
	"os"
	"testing"

	"github.com/limidus/kdb/go/kdb/metrics"
	"github.com/limidus/kdb/go/kdb/storage"
	"github.com/limidus/kdb/go/kdb/storage/engine"
	"github.com/limidus/kdb/go/kdb/storage/io"
	"github.com/limidus/kdb/go/kdb/storage/wal"
)

func newDiskBackedServerEngine(b *testing.B) *engine.ServerEngine {
	b.Helper()
	root, err := os.MkdirTemp("", "kdb-bench-wal")
	if err != nil {
		b.Fatal(err)
	}
	b.Cleanup(func() { os.RemoveAll(root) })

	store, err := io.NewOSByteStore(io.PlatformIOConfig{RootDirectory: &root})
	if err != nil {
		b.Fatal(err)
	}
	shim := io.NewFileBackedPlatformIO(io.PlatformIOConfig{RootDirectory: &root}, store)
	cfg := storage.StorageEngineConfig{GlobalMemoryBudgetBytes: 64 << 20, IOShim: shim}

	w, err := (&wal.DefaultFactory{}).OpenOrCreate("bench-ns", cfg, shim)
	if err != nil {
		b.Fatal(err)
	}
	b.Cleanup(func() { _ = w.Close() })

	return engine.NewServerEngine("bench-ns", cfg, w)
}

// BenchmarkWriteBlobConcurrent_DiskWAL is the Phase 0 baseline for the
// single most important bottleneck: ServerEngine.WriteBlob currently
// holds one engine-wide mutex across a synchronous WAL fsync
// (server_engine.go), so throughput here is expected to be bounded by
// disk fsync latency divided by exactly one concurrent writer at a time,
// regardless of -cpu.
func BenchmarkWriteBlobConcurrent_DiskWAL(b *testing.B) {
	for _, parallelism := range []int{1, 8, 64, 256} {
		b.Run(fmt.Sprintf("parallel-%d", parallelism), func(b *testing.B) {
			e := newDiskBackedServerEngine(b)
			metrics.Default.Reset()
			b.SetParallelism(parallelism)
			b.ReportAllocs()
			b.ResetTimer()
			b.RunParallel(func(pb *testing.PB) {
				payload := make([]byte, 128)
				i := 0
				for pb.Next() {
					payload[0] = byte(i)
					if _, err := e.WriteBlob(payload); err != nil {
						b.Fatal(err)
					}
					i++
				}
			})
			b.StopTimer()
			reportStages(b)
		})
	}
}

func reportStages(b *testing.B) {
	b.Helper()
	for _, s := range metrics.Default.Snapshot() {
		b.Logf("stage=%-14s count=%-8d mean=%-12s p50=%-12s p99=%-12s max=%s",
			s.Stage, s.Count, s.Mean, s.P50, s.P99, s.Max)
	}
}
